src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -614,6 +614,59 @@ internal static void RegisterMemoryLoadChangeNotification(float lowMemoryPercent
             }
         }
 
+        [Flags]
+        internal enum NoGCRegionCallbackWork
+        {
+            None,
+            Execute,
+            Detach
+        }
+
+        internal enum EnableNoGCRegionCallbackStatus
+        {
+            Succeed,
+            InsufficientBudget
+        }
+
+        private static Action? _NoGCCallback;
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NoGCRegionCallback(NoGCRegionCallbackWork work)
+        {
+            Action? callback = _NoGCCallback;
+            if (work.HasFlag(NoGCRegionCallbackWork.Detach))
+            {
+                _NoGCCallback = null;
+            }
+            if (work.HasFlag(NoGCRegionCallbackWork.Execute))
+            {
+                callback?.Invoke();
+            }
+        }
```

Comment:
@AaronRobinsonMSFT is this the recommended way to do this sort thing? we have a callback that we want to call on the finalizer thread. is it better to do it this way or just call the callback directly on the native code side?

Reply:
We keep adding more and more of one-off checks for the extra work to execute on finalizer thread. It would be better to come up with a more pay-for-play scheme. 

- Introduce a static delegate that represents extra work to do on finalizer thread. This delegate should be null most of the time, so the only fixed overhead that we pay is a simple null check (no calls, etc.)
- When some part of the system wants to do extra work on the finalizer thread, it should add its work to this delegate and remove its work from this delegate once it is no longer needed. The no-GC callback should be the first use case for this delegate. 
- In separate PR, we can convert other ad-hoc schemes that we use to schedule work on the finalizer thread to use this static delegate as well.

Reply:
@Maoni0 Given an `Action` is involved, I think it is. Either we handle lifetime via a managed static or via a GC handle in unmanaged code. I think this is easier to audit and make correct, so looks good. If the argument had been a C# function pointer (for example, `delegate* <void>`) then I think pushing that down into the runtime would be more appropriate.

@cshung I see this was renamed to `FinalizerAction`, which I assume is in regards to @jkotas's comment about making a generalized mechanism. If this is the case, is the plan to make this file the location for that mechanism? Perhaps the majority of the Finalizer cleanup routines can be written in managed code at which point having the management logic in managed is appropriate. I don't know the answer to that at the moment though. @jkotas's second point isn't satisfied by the current implementation. Are you planning on updating this routine to address that feedback?

Reply:
> is the plan to make this file the location for that mechanism?

That was my idea - it seems reasonable to me to use `GC.CoreCLR.cs` or `GC.NativeAOT.cs` as a place for managed finalizer action. For NativeAOT, `__Finalizer.cs` was too low level for that.

> @jkotas's second point isn't satisfied by the current implementation.

I am not sure about that - the way I think about it is that when I pass a `finalizerActionKind` that is not `FinalizerActionKind.None`, I am logically adding work to that delegate. Is there another way of doing that?

Reply:
There is more than a null check to do work here. I read the suggestion as the Finalizer should have a mechanism to make a single null check for whether to consider doing any more work not specfically for this scenario. Right now, the extra work is captured in `Thread::HaveExtraWorkForFinalizer()` which is a complex predicate of conditions where all need to be evaluated before the question can be answered. The generalized mechanism would reduce that to a single null check rather than a series. Right now that isn't being done because the check in this case is captured as an enum value the VM must pass through. The two ideas should be separated.

The null/non-null check is for do or don't do work - perhaps that can't be fully done right not but I can imagine calling into managed to check for now and in the longer term that would be removed. If it is non-null there is work and each "work item" becomes some potential action/flag/delegate/function pointer etc. I don't know at present but right now the logic is represented by sharing a set of flags and by sharing additional state in `Thread::HaveExtraWorkForFinalizer()` between the VM and managed code.

Reply:
I must be missing something then. The do or don't do work decision should be made before we enter managed code, which makes this particular location a wrong site for the comment.

The `null` check is `finalizerAction != finalizer_action_kind::none`, and it is currently done twice. Once in `Thread::HaveExtraWorkForFinalizer()` and another in `FinalizerThread::FinalizerThreadWorker` right before we call back to managed code.

The check here is meant to decide which subset of work is supposed to be performed - the do or don't do decision is made already.

Reply:
> The do or don't do work decision should be made before we enter managed code,

That isn't a requirement, but it can be. I see what you are doing here, but I don't an obvious way forward for the other work items for others to follow. Can you see a way to handle the other work items using the current pattern, like `SyncBlock` cleanup?

Reply:
Reading a bit about `InteropSyncBlockInfo::FlushStandbyList`, I have no idea why we would want to move them into managed code, at best we will call that in the `FinalizerAction` and then QCall back into native code to do the work.

Suppose you have some extra managed code (for now, let's call that `ExtraManagedWork`) that you would want to perform on the finalizer thread, here is how I would do it.

Modify `FinalizerActionKind` so that it contains a new flag entry `ExecuteExtraManagedWork = 4`.

By the time we compute the value of `finalizerActionKind`, logically or the value where it makes sense. Something like this:

```c++
// We can potentially aggregate multiple requests to do work on finalizer action here
finalizer_action_kind finalizerActionKind = GCHeapUtilities::GetGCHeap()->GetNoGCFinalizerAction(/* reset = */true);
if (extraManagedWorkRequired)
{
    finalizerActionKind |= finalizer_action_kind.extra_managed_work;
}
```
A similar business needs to happen with `Thread::HasExtraWorkForFinalizer`.

Then on the managed code add the appropriate branch:

```c#
if ((finalizerActionKind & FinalizerActionKind.ExecuteNoGCCallback) != 0)
{
    PerformExtraManagedWork();
}
```

Reply:
> I have no idea why we would want to move them into managed code,

In an effort to make `Thread::HaveExtraWorkForFinalizer()` be as simple as a `null` check. The entire point of the comment was to simplify the "finalizer has extra work" scenario. If you think approached provided here is sufficient to satisfy that then we can move forward but passing around flags seems tricky given the flags seem to have transient state as opposed to either being a callback to call or not in some type of linked list or queue.

Reply:
Thanks, @AaronRobinsonMSFT. Now I got it. I was so fixated on the idea of managed/unmanaged that I was lost, I was thinking about minimizing managed/native transitions, which will have a perf impact.

I am not convinced that is the right thing to do, both for perf and code quality.

From a perf perspective, I don't think we will save much. We are reducing a few method calls to a null check, but this is done for every finalizer thread invocation, which does much more things. 

From a code quality perspective, It might appear cleaner, but we are trading off on transparency. Explicit checks will be replaced by a comparison of opaque delegate pointers. This makes it much harder to reason about what work we are going to do in what order.

Reply:
Let's come back to this style issue once the other feedback is addressed.

Reply:
This is probably irrelevant now as we no longer change the `Thread::HaveExtraWorkForFinalizer` method.